### ISSUE TITLE
Rewrite towards future goal?

### ISSUE THREAD

**Issue by gishmel** [aurelia/validation#161]
I have discussed some with Rob already and wanted to get others opinions on the matter of this library and being less tied to the framework and more pluggable. I also envision adopting a much more functional approach than the large number of classes that are currently being used for each aspect of the plugin. I hope to give approximately the same type and name of API calls but I believe some are subject to change as I don't think all the options are as necessary if we can just have a generic that we give an object. It is still up in the air and even if we go for the more dynamic functional approach we can always export the API over top of the more dynamic functional approach. Please offer up ideas and criticisms as you see fit and I will try to respond and converse as appropriate. Thanks. @EisenbergEffect @PWKad

**Comment by EisenbergEffect** [comment]
@PWKad @gishmel 

I think we will need to do some significant work here:

* Split this library into two parts: 
  * A "POJO" validation library that can work against any object and that has nothing to do with UI at all. (aurelia-validation)
  * An adapter that allows the templating engine to connect with the validation system for the purpose of displaying the errors, etc. (aurelia-templating-validation)
* Remove all hacks or dependencies on internal or semi-private APIs. I'm not sure if many exist, but I had the impression there was some code in there that is a bit hackish.
* Remove any Aurelia dependencies that aren't strictly necessary. This can be part of the splitting process. Ideally, the validation library itself would have no dependencies. A couple exceptions I can think of would be metadat and DI.

We need to try to do this ASAP in the event that we need to add new extensibility points to the templating library. Once we hit Beta 2, we are on lock down for new features, so we won't be able to do it until after v1 release if it doesn't get in by then.

**Comment by PWKad** [comment]
Agreed I think that's what I was going to chime in with as well.  The first should be completely separated from the DOM so that it's more re-usable.

I think a few of the issues brought up by others are around the over-use of promises.

**Comment by gishmel** [comment]
I very much like this idea and believe this is for the best. The library currently is difficult to reason about where it is doing certain functionality as it seemingly repetitive and needlessly large. I can create the aurelia-validation library and thats interesting I will look through the issues.

**Comment by nomack84** [comment]
One must be dependency should also be `aurelia-i18n`, currently the library is reinventing the wheel around i18n, and it should simply use `aurelia-i18n`, at least that's what I think..

**Comment by EisenbergEffect** [comment]
Yes, good point, any i18 functionality should come from our library for that purpose. It might be possible to create an "interface" so that different implementations can be plugged in.

**Comment by gishmel** [comment]
I am confused what must be a dependency of aurelia-i18?

**Comment by nomack84** [comment]
for translating the error messages

**Comment by gishmel** [comment]
So you are stating that the adapter must utilize the aurelia-i18n library or plugin?

**Comment by nomack84** [comment]
The validation library should/must support i18n for exposing error messages in the different languages, current version does that, but not in the best possible way and it doesn't use `aurelia-i18n`.

**Comment by nomack84** [comment]
Is possible that the future plugin `aurelia-templating-validation` should depend on i18n for showing the messages in the correct lang.

**Comment by gishmel** [comment]
Or its a pipeline such that aurelia-validation is the start of said pipe which feeds aurelia-i18n which then finally feeds into aurelia-templating-validation adapter. But sure it is possible to architect in many different ways so not sure which would be most desirable.

**Comment by nomack84** [comment]
well, for the most desirable you should ask @EisenbergEffect or @PWKad :joy:

**Comment by jods4** [comment]
Is there a plan or anything?

I wanted to give aurelia-validation some refactoring love, but my time is limited and I certainly don't want to invest efforts into something that is going to be thrown away anyway.

aurelia-validation is currently doing lots of things in a so-so way, off the top of my head:
- overuse of promises, poor async handling
- the way `threshold` works for all validators is not good design
- the way validation rules are defined is unintuitive and not as simple as it should be
- there should be synchronous validation apis for the 90% of validations that actually are synchronous (open to discussion: should we even support async validations out of the box?)
- the way messages are formatted and i18n is not great + it's something that should be shared across aurelia
- the way paths are observed should be shared from aurelia. `PathObserver` belongs to `aurelia-binding`, not `aurelia-validation`
- Does a Debouncer belong to this project? It seems to me there has to be another one somewhere in aurelia
- The way the `validation` binding works has some (possibly hard) issues. Not sure how to fix this.
- Should aurelia provide a ViewStrategy that is not bound to Bootstrap by default? I'm not using bootstrap and I'm using my own, but it seems to me that rather than Bootstrap a simple strategy that adds a class should be the built-in default.
- and more...

Also, if we start thinking of creating an aurelia-free, functional, validation library, complemented by an aurelia adapter, than maybe we should re-use something existing. There are lots of validation functions already coded in the open-source world.

**Comment by EisenbergEffect** [comment]
We don't want to throw things away. There's been a lot of work done. But, it does need some significant re-factoring and revisions. We want to gather some feedback and have some discussion here so we can see the extent that we need to go to and the nature of the changes. I've got some things that are critical from my perspective, but I know the community has some good feedback, so we want to hear that.

**Comment by jods4** [comment]
@EisenbergEffect sorry, I expressed myself poorly. 
My point was: I know aurelia-validation is not going away; I want to be sure that if I contribute some work to make aurelia-validation better it will not be useless.

**Comment by gishmel** [comment]
I think the library as far as the validation work is concerned is trivial to implement and the more important matter is how to connect this plugin with aurelia which is through the described aurelia-templating-validation. Also I don't follow when you state their are validation functions written in the open source world? Are you saying we should use an existent library? Cause from my vantage point we should just use javascript and its a matter of a few higher order functions working in concert to allow for extensible and robust behavior that is similar to the existent library but fewer options and more configurable. Thats just my opinion.

**Comment by jods4** [comment]
@gishmel I do not actually believe that we should use existing validation libraries but it was a comment referring to this from @EisenbergEffect:
> Split this library into two parts: 
A "POJO" validation library that can work against any object and that has nothing to do with UI at all. (aurelia-validation)
An adapter that allows the templating engine to connect with the validation system for the purpose of displaying the errors, etc. (aurelia-templating-validation)

If `aurelia-validation` is not actually tied to Aurelia directly, then I think we could reuse js code that already exists. We are not the first to tackle validation.

But then in my opinion even the splitted `aurelia-validation` is rather tied to Aurelia, for localization or to be able to react automatically to property changes.

**Comment by gishmel** [comment]
No its not tied to the property changes that is the responsibility of the aurelia-templating-validation library to recall our plugin/library and we can give them the validation output back.

**Comment by jods4** [comment]
@gishmel makes sense, although the api needs to be well designed for it to work (consider `<button enabled.bind="model.isValid">`).

There is the i18n question left, then my initial point was: if the code is a generic, aurelia-independent, validation library, then maybe we should just use something that already exists.

**Comment by gishmel** [comment]
Such as what library?

**Comment by jods4** [comment]
I am not sure, there are tons of JS validation libraries. 

Of varying quality, with more or less dependencies, more or less tied to UI and a set of rules that can more or less easily be re-used.

There's a lot to look at: parsley.js, validator.js, verify.js, jQuery.validate, knockout.validation, chriso/validator and a ton more if you use Google I guess.

**Comment by gishmel** [comment]
Yeah I did google and get those same libraries looked at a few. Still think we need nothing but a bit of ES6 and good discipline and we would have all their features and functionality with none of the overhead but thats once again my opinion. Which I have overly shared and look forward to seeing the overhaul plan at some point. Thanks for your time.

**Comment by PWKad** [comment]
@gishmel Sorry I didn't chime in much earlier.

There is a lot of work to be done here to refactor the DOM manipulation out in to a pluggable fashion so that the aurelia-templating-validation can be created.  It sounds like you have some good ideas.  How can we turn those ideas in to some actionable items?

@EisenbergEffect listed a few starting points above, @jods4 listed some more detailed, and the issue I linked re-affirms that the main point should be to remove the templating portions ASAP.

I think a good starting point is to start a branch removing *all* DOM manipulation and use that as a basis for the refactor.  With those pieces removed perhaps the next step is to identify the current work flow that the validation library goes through to validate a property and then identify an improved workflow.  Then as we progress through the refactoring we can start to simplify and remove dead code.  My opinion on this is that I think there are some additional properties that exist currently we can probably reduce down.  Based on the feedback we've seen so far though the top priority should be getting the two pieces split-up so perhaps if we can decide on any changes in the workflow we can divvy the work up.

Thoughts?

**Comment by PWKad** [comment]
@gishmel @jods4 @nomack84 @EisenbergEffect 

I've moved this discussion in to a task list on a separate issue since I don't have the ability to edit the first post to keep it here.  I've took all of the discussion points that I saw and could move over in to that issue.  I'm going to close this for now but I would invite all of you to please take a look at that list and make sure I didn't miss anything and it all looks good.

If so I'm going to start moving issues over to the new repository and opening issues for each of the tasks.  If anyone feels strongly about taking on some of the tasks just let me know which ones once we get them created.
